streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -338,6 +338,23 @@ public void setStateListener(final KafkaStreams.StateListener listener) {
         }
     }
 
+    /**
+     * An app can set a single {@link StreamThread.StateListener} so that the app is notified when state changes.
+     *
+     * @param listener a new StreamThread state listener
+     * @throws IllegalStateException if this {@code KafkaStreams} instance is not in state {@link State#CREATED CREATED}.
+     */
+    public void setStreamThreadStateListener(StreamThread.StateListener listener) {
```

Comment:
If we are adding this function in `KafkaStreams` then we do need a KIP.. but I was thinking if we can just add this in `StreamThread`, which is an internal class and hence doing so does not need a KIP. We can, instead, add a `KafkaStreamsWrapper` (it is similar to `TopologyWrapper` and `TopologyTestDriverWrapper` to allow unit test code to access their internal private fields) which can access the threads and globalThread, and then adds this function in this class to manipulate them.

Also note that StreamThread already has a state change listener `final class StreamStateListener implements StreamThread.StateListener` so our `StateListenerStub` should not completely replace its logic. Instead we can extend that listener to `StateListenerStub` which does the state change tracing in additional to the necessary logic.

Reply:
Please confirm: threads and state fields of KafkaStreams can be changed to protected.
Otherwise the new method in KafkaStreamsWrapper still cannot access them.

Reply:
Yes, they can.